platforms/android/tangram/src/main/java/com/mapzen/tangram/MapView.java

```diff
@@ -18,45 +19,130 @@
-     * the callback will be made on the UI thread
+     * @return An instance of {@link MapController} if previously created else, null
      */
-    @NonNull
-    public MapController getMap(@Nullable final MapController.SceneLoadListener listener, @Nullable final HttpHandler handler) {
-        if (mapController != null) {
-            return mapController;
+    @Nullable
+    public MapController getMapAsync(@Nullable final MapReadyCallback readyCallback,
+                            @Nullable final MapController.SceneLoadListener sceneLoadListener,
+                            @Nullable final HttpHandler handler) {
+
+        synchronized (lock) {
+            if (mapController != null) {
+                return mapController;
+            }
+        }
+
+        disposeMapReadyTask();
+        if (glSurfaceView == null) {
+            glSurfaceView = new GLSurfaceView(getContext());
         }
-        mapController = getMapInstance();
-        mapController.setSceneLoadListener(listener);
-        mapController.init(handler);
+        final Context ctx = glSurfaceView.getContext();
+
+        getMapAsyncTask = new AsyncTask<Void, Void, MapController>() {
+            @Override
+            protected MapController doInBackground(Void... voids) {
+                synchronized (lock) {
+                    if (mapController != null) {
+                        return mapController;
+                    }
+                }
+                MapController controller = getMapInstance();
+                controller.init(ctx, handler);
+
+                return controller;
+            }
+
+            @Override
+            protected void onPostExecute(MapController controller) {
+                synchronized (lock) {
+                    if (mapController != null) {
+                        mapController.dispose();
+                    }
+                    mapController = controller;
+                    configureGLSurfaceView();
+                    mapController.postInit();
+                    mapController.setSceneLoadListener(sceneLoadListener);
+                    readyCallback.onMapReady(mapController);
+                }
+            }
+
+            @Override
+            protected void onCancelled(MapController controller) {
+                if (controller != null) {
+                    controller.dispose();
+                }
+            }
+        }.execute();
+
+        return null;
+    }
 
-        return mapController;
+    /**
+     * Returns the mapController previously created using
+     * {@link MapView#getMapAsync(MapReadyCallback)} or {@link MapView#getMapAsync(MapReadyCallback, MapController.SceneLoadListener)} or
+     * {@link MapView#getMapAsync(MapReadyCallback, MapController.SceneLoadListener, HttpHandler)}
+     * @return An instance of {@link MapController} if previously created else, null
+     */
+    @Nullable
+    public MapController getMapController() {
```

Comment:
I don't think I understand the need for this method. It seems like an application could keep its own nullable reference to a `MapController` for this purpose, right?

Reply:
Yeah, updated this.